File: JavaScript/leetcode75/src/main/Problem28_Dota2Senate.js
/**
 * @param {string} senate
 * @return {string}
 */
var predictPartyVictory = function (senate) {
  let rQueue = [];
  let dQueue = [];
  for (let i = 0; i < senate.length; i++) {
    if (senate[i] == "R") {
      rQueue.push(i);
    } else {
      dQueue.push(i);
    }
  }
  let n = senate.length;
  while (rQueue.length > 0 && dQueue.length > 0) {
    let rturn = rQueue.shift();
    let dturn = dQueue.shift();
    if (dturn < rturn) {
      dQueue.push(dturn + n);
    } else {
      rQueue.push(rturn + n);
    }
  }
  return rQueue.length > 0 ? "Radiant" : "Dire";
};

module.exports = predictPartyVictory;

File: Java/leetcode75/src/main/java/org/phoenix/leetcode/challenges/Problem28_Dota2Senate.java
package org.phoenix.leetcode.challenges;

import java.util.LinkedList;
import java.util.Queue;

public class Problem28_Dota2Senate {
    public String predictPartyVictory(String senate) {
        Queue<Integer> rQueue = new LinkedList<>();
        Queue<Integer> dQueue = new LinkedList<>();
        for (int i = 0; i < senate.length(); i++) {
            if (senate.charAt(i) == 'R') {
                rQueue.offer(i);
            } else {
                dQueue.offer(i);
            }
        }
        int n = senate.length();
        while (!rQueue.isEmpty() && !dQueue.isEmpty()) {
            int rturn = rQueue.poll();
            int dturn = dQueue.poll();
            if (dturn < rturn) {
                dQueue.add(dturn + n);
            } else {
                rQueue.add(rturn + n);
            }
        }
        return dQueue.isEmpty() ? "Radiant" : "Dire";
    }
}
